walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,144 @@
+                        -> event.getAttendees().stream().allMatch(
+                            attendee -> optionalAttendees.contains(attendee)));
+    // Change this to get overlapping events.
+    ArrayList<TimeRange> timesWithOptional = getUnavailableTimes(eventList);
+    ArrayList<TimeRange> timesWithoutOptional = getUnavailableTimes(noOptional);
+    toReturn = getInverse(timesWithOptional, request);
+    if (toReturn.isEmpty() && !timesWithoutOptional.isEmpty()) {
+      toReturn = getInverse(timesWithoutOptional, request);
+    }
+
+    return toReturn;
+  }
+
+  private int getMinStartTime(int currentStartTime, ListIterator<Event> it) {
+    int minStartTime = currentStartTime;
+    while (it.hasNext()) {
+      Event nextEvent = it.next();
+      TimeRange nextEventTime = nextEvent.getWhen();
+      // If an event is in the current range.
+      if (nextEventTime.end() >= currentStartTime) {
+        minStartTime = Math.min(minStartTime, nextEventTime.start());
+      } else {
+        // In this case it is not possible to expand this event any farther,
+        // the iterator already points to the next element we should continue
+        // from.
+        // Reset the iterator so the last tried value isn't skipped.
+        it.previous();
+        break;
+      }
+    }
+    return minStartTime;
+  }
+  // This function assumes that the arguments it is given are in decreasing
+  // order. It thus parses them from end to beginning.
+  private ArrayList<TimeRange> getInverse(ArrayList<TimeRange> fullTimes,
+                                          MeetingRequest request) {
+    long meetingDuration = request.getDuration();
+    ArrayList<TimeRange> toReturn = new ArrayList<>();
+    int priorStartTime = TimeRange.START_OF_DAY;
+    // Finally we loop through the time in between these events (the inverse
+    // of the set), and if these times are within the length of the meeting
+    // we add them to our answer (going from smallest to biggest).
+
+    for (ListIterator<TimeRange> it = fullTimes.listIterator(fullTimes.size());
+         it.hasPrevious();) {
+      TimeRange currentRange = it.previous();
+      int currentStartTime = currentRange.start();
+      int currentEndTime = currentRange.end();
+      int currentDuration = currentStartTime - priorStartTime;
+
+      if (currentDuration >= meetingDuration) {
+        // The meeting is long enough, only take an action in this case.
+        toReturn.add(
+            TimeRange.fromStartDuration(priorStartTime, currentDuration));
+      }
+      priorStartTime = currentEndTime;
+    }
+    // Also need to make sure to include the last range as well.
+    int endDuration = TimeRange.END_OF_DAY - priorStartTime;
+    if (endDuration >= meetingDuration) {
+      toReturn.add(
+          TimeRange.fromStartEnd(priorStartTime, TimeRange.END_OF_DAY, true));
+    }
+    return toReturn;
+  }
+
+  private ArrayList<TimeRange> getUnavailableTimes(ArrayList<Event> eventList) {
```

Comment:
Could you add function description? Does it merged the overlapped time range of all events? I feel you don't need to do it specifically, for the given event list, sort it by start time in asc order, 
set event_end_time = START_OF_DAY,
then traverse through each event, if the current event start time larger than the event_end_time, you get a available slot, check the length to decide whether add it to your result list, otherwise, update your event_end_time = max(event_end_time, event.end_time).